packages/tupaia-web/src/utils/useDateRanges.ts

```diff
@@ -0,0 +1,150 @@
+import { useSearchParams } from 'react-router-dom';
+import moment, { Moment } from 'moment';
+import {
+  GRANULARITY_CONFIG,
+  GRANULARITIES,
+  getDefaultDates,
+  getLimits,
+  momentToDateDisplayString,
+  GRANULARITIES_WITH_ONE_DATE,
+  roundStartEndDates,
+} from '@tupaia/utils';
+import { DashboardItemType, SingleMapOverlayItem } from '../types';
+import { DEFAULT_PERIOD_PARAM_STRING } from '../constants';
+
+// converts the date range to a URL period string
+const convertDateRangeToUrlPeriodString = (
+  {
+    startDate,
+    endDate,
+  }: {
+    startDate: Moment;
+    endDate: Moment;
+  },
+  granularity = GRANULARITIES.DAY,
+) => {
+  if (!(startDate || endDate)) return null;
+
+  const { urlFormat } = GRANULARITY_CONFIG[granularity];
+
+  const formattedStartDate = momentToDateDisplayString(
+    startDate,
+    granularity,
+    urlFormat,
+    undefined,
+  );
+  const formattedEndDate = momentToDateDisplayString(endDate, granularity, urlFormat, undefined);
+
+  return GRANULARITIES_WITH_ONE_DATE.includes(granularity)
+    ? formattedEndDate
+    : `${formattedStartDate}-${formattedEndDate}`;
+};
+
+// converts the URL period string to a date range
+const convertUrlPeriodStringToDateRange = (
+  periodString: string,
+  granularity = GRANULARITIES.DAY,
+): {
+  startDate: Moment | null;
+  endDate: Moment | null;
+} => {
+  const [startDate, endDate] = periodString.split('-');
+
+  if (!startDate || periodString === DEFAULT_PERIOD_PARAM_STRING) {
+    return {
+      startDate: null,
+      endDate: null,
+    };
+  }
+
+  const { urlFormat } = GRANULARITY_CONFIG[granularity];
+
+  const momentStartDate = moment(startDate, urlFormat);
+
+  if (GRANULARITIES_WITH_ONE_DATE.includes(granularity)) {
+    return {
+      startDate: momentStartDate,
+      endDate: momentStartDate,
+    };
+  }
+  // We rely on dates being rounded in state for range formats
+  const momentEndDate = moment(endDate, urlFormat);
+  return roundStartEndDates(granularity, momentStartDate, momentEndDate);
+};
+
+/**
+ * This handles the logic for setting and getting the dates for the dates with map overlays and reports, including updating the URLSearchParams
+ */
+export const useDateRanges = (
```

Comment:
This hook is looking great. To me it's a much more easy to understand place to keep the logic for the date pickers. The code is really dense but there is no easy way to avoid that, right. I assume most of it was copied across from web-frontend which is cool. I was just about to ask if you could copy across and update the dateRangePicker tests but then I noticed that there are some in ui-components for the dateRangePicker so I think that should be fine.